src/Http/Http.Abstractions/src/Extensions/UsePathBaseMiddleware.cs

```diff
@@ -41,37 +41,36 @@ public UsePathBaseMiddleware(RequestDelegate next, PathString pathBase)
         /// </summary>
         /// <param name="context">The <see cref="HttpContext"/> for the current request.</param>
         /// <returns>A task that represents the execution of this middleware.</returns>
-        public async Task Invoke(HttpContext context)
+        public Task Invoke(HttpContext context)
         {
             if (context == null)
             {
                 throw new ArgumentNullException(nameof(context));
             }
 
-            PathString matchedPath;
-            PathString remainingPath;
-
-            if (context.Request.Path.StartsWithSegments(_pathBase, out matchedPath, out remainingPath))
+            if (context.Request.Path.StartsWithSegments(_pathBase, out var matchedPath, out var remainingPath))
             {
-                var originalPath = context.Request.Path;
-                var originalPathBase = context.Request.PathBase;
-                context.Request.Path = remainingPath;
-                context.Request.PathBase = originalPathBase.Add(matchedPath);
-
-                try
-                {
-                    await _next(context);
-                }
-                finally
-                {
-                    context.Request.Path = originalPath;
-                    context.Request.PathBase = originalPathBase;
-                }
+                return InvokeCore(context, matchedPath, remainingPath);
             }
-            else
+            return _next(context);
+        }
+
+        private async Task InvokeCore(HttpContext context, string matchedPath, string remainingPath)
```

Comment:
In Kestrel we often use the "Awaited" suffix to indicate why we're using this patter. So this would be `InvokeAwaited`. I'm okay with taking this as-is though.